### PR TITLE
Replace invoking self.value if there is a user defined init, avoiding arbitrary code execution

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117818

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: test/dynamo/test_repros.py [hunk 1]
```diff
@@ -4000,6 +4000,74 @@ def func3(x, y):
...
+            ucm = UserCtxManager(y)
+            return x * y[0]
+
+        cnt = torch._dynamo.testing.CompileCounter()
+        opt_fn = torch._dynamo.optimize(cnt, nopython=True)(fn)
+        x = torch.rand([2, 2])
+        self.assertEqual(opt_fn(x, [5]), fn(x, [5]))
+        self.assertEqual(cnt.frame_count, 1)
+
+    def test_user_ctor_ctx_manager_custom_init_graph_break(self):
```
Comment (voznesenskym) [inline test/dynamo/test_repros.py:4042]: self nit - crappy name

**Comment by voznesenskym** [comment]
Most of the test failures are unexpected success, which is nice, I guess, but main is pretty flaky rn so I am going to wait

File: torch/_dynamo/variables/user_defined.py [hunk 2]
```diff
@@ -291,9 +291,16 @@ def call_function(
...
-                unwrapped_args,
-                cm_obj=self.value(*unwrapped_args),
+            if self.value.__init__ == object.__init__:
+                return GenericContextWrappingVariable(
+                    unwrapped_args,
+                    cm_obj=self.value(*unwrapped_args),
+                )
+
+            return tx.inline_user_function_return(
+                variables.UserFunctionVariable(self.value.__init__, source=self.source),
```
Comment (ezyang) [inline torch/_dynamo/variables/user_defined.py:301]: Why is this right? `__init__` doesn't even return anything typically

Comment (voznesenskym) [reply torch/_dynamo/variables/user_defined.py:301]: It produces a ConstantVariable(NoneType), which iirc, is the right the right thin to push onto the stack.

File: torch/_dynamo/variables/user_defined.py [hunk 3]
```diff
@@ -291,9 +291,22 @@ def call_function(
...
+            if self.value.__init__ == object.__init__:
+                return GenericContextWrappingVariable(
+                    unwrapped_args,
+                    cm_obj=self.value(*unwrapped_args),
+                )
+
+            try:
+                fn = variables.UserFunctionVariable(
+                    self.value.__init__, source=self.source
+                )
```
Comment (mlazos) [inline torch/_dynamo/variables/user_defined.py:303]: It seems weird to me to construct a UserFunctionVariable directly, shouldn't we call builder on the type of value, and then call_function from there to handle the cases in `UserDefinedClass`?

At the very least, we should guard here unless the guards were already taken care of when this var was built.

File: torch/_dynamo/variables/user_defined.py [hunk 4]
```diff
@@ -291,9 +291,22 @@ def call_function(
...
+            except AssertionError:
+                unimplemented("Failed to get __init__ function of %s" % self.value)
+            return tx.inline_user_function_return(
+                fn,
```
Comment (mlazos) [inline torch/_dynamo/variables/user_defined.py:307]: Hmm here we're returning the traced return value of `__init__`, don't we need to create a `GenericContextWrappingVariable` to return?

**Review by ezyang — APPROVED** [review]
New version seems fine as a bandaid

**Comment by voznesenskym** [comment]
suffering from success lol, have to keep removing unexpected success tests from failure lists! Pinged @zou3519 to see if there's a faster way...

**Pull request closed without merging**